### DISCUSSION TITLE
Plugin architecture

### DISCUSSION THREAD

**Discussion by fadwamahmoud** [librocco/librocco#323] [General]
I'm still working on this and I'm leaving the diagram here for reference. This is a high level view of how the plugin system should work.

We have 3 components:
1. Main app: 
 - Calls pluginManager.getBook()
 - Provides a settings page for adding/removing plugins
 
2. Plugin manager extension:
 - Receives messages from client and communicates them to background scrip 
 - background script fetches the required resources
 - scraping code is executed on those resources
 - communicates books back to client app
 
3. scraper scripts: 
 - in a separate repo

The idea is that basically the user will have the ability to paste script code to create plugins. Then they can do that plus control which ones to get called from the plugin manager extension UI.

This diagram is a WIP
https://app.eraser.io/workspace/oeZ2EBXUjUVHeCW4i8cm?origin=share

![diagram-export-8_8_2023, 12 30 27 AM](https://github.com/librocco/librocco/assets/41344489/e90faf19-32bc-4453-963b-98842ebbd8ca)

**Comment by fadwamahmoud** [comment]
Next steps:
1. Integrate pluginManager with plugins
2. Create a separate package for the plugin manager
3. Remove scraping code from extension repo

**Comment by ikusteu** [comment]
I'm wary of using the term **plugin manager** for what is essentially a **book data extension**. 

I think the pluggable api is a huge chunk of the design of the entire platform. And I would like to balance it between not dwelling too much on the what-ifs, but making it so that we lay down the good foundation in case we want to extend it at some point.

@fadwamahmoud, first and foremost, while I'm wary of the terminology, I don't have an objection on the architecture of the extension and the book fetching logic overview 👍 

I would, though, like to give some ideas on the way the api should look like and how the existing app should integrate the pluggable api.

## 1. Create a `BookFetcherPlugin`

The plugin interface should be just that: an interface. It should specify some required methods and some optional ones. For starters, we need a method to fetch the books and at this point, I don't think we need much more than that:
```ts
interface BookFetcherPlugin {
  fetchBooks(isbns: string[]): Promise<BookEntry[]>
}
```

## 2. Create a flow to register that plugin and a fallback

I've imagined it something like this:
- there's a `plugin` method on the db interface which accepts a string (at this point it should only accept `"book-fetcher"`). The method would return a `BookFetcherPlugin` with some standard (plugin core) methods, i.e.:
```ts
type LibroccoPlugin<T> = {
  register(plugin: T): LibroccoPlugin<T> // returns itself
} & T
```
Before we register the plugin, we could do `db.plugin("book-fetcher")` and that would return a fallback. The fallback should satisfy the plugin interface (in case of book fetcher, should satisfy the `BookFetcherPlugin` interface), providing us with core plugin method(s) (currently only `register`) and no-op for the rest of the plugin interface (only `fetchBooks` at this point which, in this capacity, would return `[]`). 
The core plugin methods (e.g. `register`) should always work in the same way, so that we can use it to, for instance, register the new plugin, e.g.:
```ts
// Before any custom plugin has been registered for book fetching logic
const plugin: LibroccoPlugin<BookFetcherPlugin> = db.plugin("book-fetcher") // Returns a fallback

plugin.fetchBooks(["1234567890", "1111111111"]) // Returns Promise<[]> (essentially a noop)

// We instantiate an object that satisfies the interface
const bookFetcher = new BookDataExtension() // BookDataExtension implements LibroccoPlugin<BookFetcherPlugin>

// and we register the plugin
db.plugin("book-fetcher").register(bookFetcher)

// We call the method on the plugin
db.plugin("book-fetcher").fetchBooks(["1234567890", "1111111111"]) // Returns Promise<[book1, book2]>

// If want to switch up the plugin, we simply do
const bookFetcher2 = new SomeOtherBookFetcher()
db.plugin("book-fetcher").register(bookFetcher2)
```

## 3. Implement the book data extension plugin
I would create a package with this path `pkg/plugins/book-data-extension`. There I would simply implement the function to fetch the books using the extension, i.e.:

```ts
import { BookFetcherPlugin, BookEntry } from "@librocco/shared"

const newBookFetcherPlugin = (): LibroccoPlugin => {
  const fetchBooks = async (isbns: string[]): Promise<BookEntry[]> {
    // Check if the extension is registered (more below), if not resolve to []

    // Post message to the extension

    // Listen for a response and resolve with a value
  }
 
  return { fetchBooks }
}
```

We should create a way to signal that the extension is instantiated and available for usage. I think we could simply do `window.bookDataExtension = true` and perform a check that way.

The actual functionality of the extension (current functionality) should be kept in another repo (as agreed). As a first pass, the aforementioned implementation will do, but as we progress, we could build up on the extension implementation (to end up with the result similar to one Fadwa has described above). That part of the implementation should be kept in the `pkg/plugins/book-data-extension` at which point we should probably split the "plugin" code (an object satisfying the interface) and extension code, the "meat" of the plugin, being content scripts (listening for messages and responding back) and background (doing fetching logic, registering new fetching logic, managing fetching scripts, etc.).

## Reasoning

I've thought long and hard about this and my reasoning is this:

@silviot had, at some point, mentioned that the idea for Librocco, other than being a stock management tool, is to be more like a CRM software, so my assumption is that, at some point in the future, we would need to support more pluggable api for other sections of the app. I believe this is a good foundation, allowing us to focus on the functionality at hand, but one that can easily be used in a more generic way. This way we can, for instance, implement a UI segment, which would allow us to manage multiple plugins (not just the  book data) and as such would require us to extend the plugin API: both the basic one (`LibroccoPlugin`) as well as use-case specific (`BookFetcherPlugin`, etc.). The good thing is: we don't have to do it and can focus on simple functionality right now, but if and when we decide we need it, we can easily extend the current implementation.

As for registering of plugins: 
@fadwamahmoud has originally passed the plugin instance into the `.books()` method, which can become tedious as we'd need to pass the same instance each time we call to `.books()` method. 
My idea was to pass the plugin instance to the db on initialisation and have it passed through internally.
However, in tests, we receive the db interface, not the db initialisation function as some setup is done in the background of each test. This way we can opt-in and pass a plugin instance for inside the tests when (and only when) we need to.

**Comment by Charlotte-br560** [comment]
Your outlined plugin system shows a clear communication flow between the main app, plugin manager extension, and scraper scripts. Allowing users to create plugins via script code is a powerful feature, but it ensures safety measures are in place. Providing a settings page for plugin control in the main app is user-friendly. Keeping scraper scripts in a separate repository is a good practice for versioning and maintenance. Continuously update your diagram as the system evolves. Overall, your approach looks well-structured and promising. Good luck with your project!
